### PR TITLE
Spark proposals 04

### PR DESCRIPTION
**Pull request by Blady-Com**

### CODE REVIEW

File: kernel/core/os.adb [hunk 1]
```diff
@@ -481,8 +481,9 @@ is
                 Input => (os_task_ro, os_task_current)),
       Pre => os_ghost_task_list_is_well_formed and
              os_ghost_current_task_is_ready,
+             os_ghost_task_mbx_are_well_formed (dest_id),
```
Comment (jcdubois) [inline kernel/core/os.adb]: os_ghost_task_mbx_are_well_formed is part of os_ghost_task_list_is_well_formed.
Beside there is a logical operator missing (it does not compile).

File: kernel/core/include/os_arch.ads [hunk 2]
```diff
@@ -36,7 +36,8 @@ package os_arch is
    pragma Import (C, os_arch_interrupt_is_pending, "os_arch_interrupt_is_pending");
 
    procedure os_arch_idle
-      with Global => null;
+      with Global => null,
+           Post   => os.os_ghost_task_list_is_well_formed;
```
Comment (jcdubois) [inline kernel/core/include/os_arch.ads:40]: with this the program fails to link. Not sure why.

File: kernel/core/os.adb [hunk 3]
```diff
@@ -88,9 +88,9 @@ is
    is (os_task_ro (task_id).mbx_permission);
 
    -------------------------------
-   -- os_mbx_get_mbx_permission --
+   -- os_get_task_priority --
    -------------------------------
-   --  Get the mbx permission for a given task
+   --  Get the mbx priority for a given task
```
Comment (jcdubois) [inline kernel/core/os.adb:93]: OK

File: kernel/core/os.adb [hunk 4]
```diff
@@ -224,12 +224,12 @@ is
       mbx_msg : os_mbx_msg_t)
    with
       Global => (In_Out => os_task_rw),
-      Pre => (os_ghost_task_mbx_are_well_formed (dest_id) and then
-              os_mbx_get_mbx_count (dest_id) < os_mbx_count_t'Last),
+      Pre => (os_ghost_task_mbx_are_well_formed (dest_id) and
+              not os_mbx_is_full (dest_id)),
```
Comment (jcdubois) [inline kernel/core/os.adb:228]: OK

File: kernel/core/include/os_arch.ads [hunk 5]
```diff
@@ -36,7 +36,8 @@ package os_arch is
    pragma Import (C, os_arch_interrupt_is_pending, "os_arch_interrupt_is_pending");
 
    procedure os_arch_idle
-      with Global => null;
+      with Global => null,
```
Comment (jcdubois) [inline kernel/core/include/os_arch.ads:39]: Doesn't Global => null means that the function has no side effect and therefore should not have any effect on the task list?

File: kernel/core/os.adb [hunk 6]
```diff
@@ -591,10 +592,10 @@ is
                 (task_id, mbx_index))))) /= 0)
    with
       Pre => not os_mbx_is_empty (task_id)
-             and then
+             and
```
Comment (jcdubois) [inline kernel/core/os.adb:595]: and then seems legetimate here.
I check mbx slots only if the mbx fifo is not empty.

File: kernel/core/os.adb [hunk 7]
```diff
@@ -591,10 +592,10 @@ is
...
                 (for some index in 0 .. (os_mbx_get_mbx_count (task_id) - 1) =>
                           (os_mbx_get_mbx_head (task_id) + index) = mbx_index)
-             and then os_task_rw (task_id).mbx.mbx_array
+             and os_task_rw (task_id).mbx.mbx_array
```
Comment (jcdubois) [inline kernel/core/os.adb:598]: here also, I think this should be tested only if the mbx fifo is not empty.

File: kernel/core/os.adb [hunk 8]
```diff
@@ -479,10 +479,11 @@ is
    with
       Global => (In_Out => (os_task_ready_list_head, os_task_rw, os_ghost_task_ready),
                 Input => (os_task_ro, os_task_current)),
-      Pre => os_ghost_task_list_is_well_formed and then
+      Pre => os_ghost_task_list_is_well_formed and
              os_ghost_current_task_is_ready,
-      Post => os_ghost_task_list_is_well_formed and then
-              os_ghost_current_task_is_ready
+             os_ghost_task_mbx_are_well_formed (dest_id),
```
Comment (jcdubois) [inline kernel/core/os.adb]: os_ghost_task_mbx_are_well_formed is part of os_ghost_task_list_is_well_formed. And this make the compilation failing.

File: kernel/core/os.adb [hunk 9]
```diff
@@ -610,7 +611,7 @@ is
       (not
          (for some next_id in os_task_rw'Range =>
             os_task_rw (next_id).next = task_id
-            and then
+            and
```
Comment (jcdubois) [inline kernel/core/os.adb:614]: next_id2 depends on next_id. There is a relationship between these 2 variables.

File: kernel/core/os.adb [hunk 10]
```diff
@@ -625,7 +626,7 @@ is
       (not
          (for some prev_id in os_task_rw'Range =>
             os_task_rw (prev_id).prev = task_id
-            and then
+            and
```
Comment (jcdubois) [inline kernel/core/os.adb:629]: idem between prev_id and prev_id2.

File: kernel/core/os.adb [hunk 11]
```diff
@@ -686,13 +687,13 @@ is
 
    function os_ghost_task_mbx_are_well_formed (task_id : os_task_id_param_t) return Boolean is
       (for all index in os_mbx_index_t'Range =>
-         (if (((os_mbx_get_mbx_count (task_id) > 0) and then
+         (if (((os_mbx_get_mbx_count (task_id) > 0) and
               (os_ghost_get_mbx_tail (task_id)
```
Comment (jcdubois) [inline kernel/core/os.adb:691]: os_ghost_get_mbx_tail is only valid if the fifo is not empty.

File: kernel/core/os.adb [hunk 12]
```diff
@@ -713,10 +714,10 @@ is
    -------------------------------------------------
 
    function os_ghost_head_list_task_has_higher_priority return Boolean is
-      (os_sched_get_current_list_head /= OS_TASK_ID_NONE and then
-       os_ghost_task_is_ready (os_sched_get_current_list_head) and then
+      (os_sched_get_current_list_head /= OS_TASK_ID_NONE and
+       os_ghost_task_is_ready (os_sched_get_current_list_head) and
```
Comment (jcdubois) [inline kernel/core/os.adb:718]: if os_sched_get_current_list_head returns OS_TASK_ID_NONE this is going to be bad.

File: kernel/core/os.adb [hunk 13]
```diff
@@ -713,10 +714,10 @@ is
...
          (for some task_id in os_task_rw'Range =>
-            os_ghost_task_is_ready (task_id) and then
+            os_ghost_task_is_ready (task_id) and
```
Comment (jcdubois) [inline kernel/core/os.adb:720]: if the task is not ready we don't care about its priority.

File: kernel/core/include/os.ads [hunk 14]
```diff
@@ -71,7 +71,7 @@ is
 
    subtype os_priority_t is types.uint8_t;
 
-   type os_mbx_msg_t is mod 2 ** OS_MBX_MSG_SZ;
+   type os_mbx_msg_t is mod 2 ** OS_MBX_MSG_SZ; -- Q: why modular type?
```
Comment (jcdubois) [inline kernel/core/include/os.ads]: not strictly necessary. But it doesn't hurt and it is an easy way to express the type as a power of 2.
In moth mbx can be 8, 16, 32 or 64 bits.

File: kernel/core/os.adb [hunk 15]
```diff
@@ -460,7 +460,7 @@ is
               mbx_mask or
               os_mbx_mask_t (Shift_Left
                 (Unsigned_32'(1),
-                 Natural (os_mbx_get_mbx_entry_sender (task_id, mbx_index))));
+                 Natural (os_mbx_get_mbx_entry_sender (task_id, mbx_index)))); -- medium: precondition might fail -> proposal: is sender_id valid should be a ghost function to help proof
```
Comment (jcdubois) [inline kernel/core/os.adb:463]: os_mbx_get_mbx_entry_sender is needed by the code here. It cannot be a ghost function.

File: kernel/core/os.adb [hunk 16]
```diff
@@ -224,12 +224,12 @@ is
...
-      Pre => (os_ghost_task_mbx_are_well_formed (dest_id) and then
-              os_mbx_get_mbx_count (dest_id) < os_mbx_count_t'Last),
+      Pre => (os_ghost_task_mbx_are_well_formed (dest_id) and
+              not os_mbx_is_full (dest_id)),
       Post => ((os_mbx_get_mbx_count (dest_id) =
-                 os_task_rw (dest_id).mbx.count'Old + 1) and then
+                 os_task_rw (dest_id).mbx.count'Old + 1) and
               (os_task_rw (dest_id).mbx.mbx_array
-                       (os_ghost_get_mbx_tail (dest_id))
+                       (os_ghost_get_mbx_tail (dest_id))  -- medium: postcondition might fail, cannot prove -> head + count is may be too hard to prove -> proposal: add a tail index in os_mbx_t
```
Comment (jcdubois) [inline kernel/core/os.adb:232]: It is a pity to have to add unnecessary things to the program to help the proof. It does not seem like the way to go.

File: kernel/core/os.adb [hunk 17]
```diff
@@ -479,10 +479,11 @@ is
...
       Global => (In_Out => (os_task_ready_list_head, os_task_rw, os_ghost_task_ready),
                 Input => (os_task_ro, os_task_current)),
-      Pre => os_ghost_task_list_is_well_formed and then
+      Pre => os_ghost_task_list_is_well_formed and
              os_ghost_current_task_is_ready,
-      Post => os_ghost_task_list_is_well_formed and then
-              os_ghost_current_task_is_ready
+             os_ghost_task_mbx_are_well_formed (dest_id),
+      Post => os_ghost_task_list_is_well_formed and
+              os_ghost_current_task_is_ready -- medium: postcondition might fail, cannot prove os_ghost_current_task_is_ready -> is it true in every cases below?
```
Comment (jcdubois) [inline kernel/core/os.adb]: Sending a mbx should not change the scheduled task. It is a non blocking call. So current should not change and still be ready.

File: kernel/core/os.adb [hunk 18]
```diff
@@ -686,13 +687,13 @@ is
 
    function os_ghost_task_mbx_are_well_formed (task_id : os_task_id_param_t) return Boolean is
       (for all index in os_mbx_index_t'Range =>
-         (if (((os_mbx_get_mbx_count (task_id) > 0) and then
+         (if (((os_mbx_get_mbx_count (task_id) > 0) and
               (os_ghost_get_mbx_tail (task_id)
-                        < os_mbx_get_mbx_head (task_id)) and then
+                        < os_mbx_get_mbx_head (task_id)) and
```
Comment (jcdubois) [inline kernel/core/os.adb:692]: mbx fifo needs to be not empty to call os_ghost_get_mbx_tail

File: kernel/core/os.adb [hunk 19]
```diff
@@ -686,13 +687,13 @@ is
...
               ((index >= os_mbx_get_mbx_head (task_id)) or
                (index <= os_ghost_get_mbx_tail (task_id))))
-          or else
-             ((os_mbx_get_mbx_count (task_id) > 0) and then
+          or
```
Comment (jcdubois) [inline kernel/core/os.adb:695]: The 2 conditions are really exclusive ...

File: kernel/core/os.adb [hunk 20]
```diff
@@ -686,13 +687,13 @@ is
...
+             ((os_mbx_get_mbx_count (task_id) > 0) and
```
Comment (jcdubois) [inline kernel/core/os.adb:696]: mbx fifo needs to be not empty to call os_ghost_get_mbx_tail

File: kernel/core/os.adb [hunk 21]
```diff
@@ -729,7 +730,7 @@ is
    function os_ghost_task_list_is_well_formed return Boolean is
       --  The mbx fifo of all tasks need to be well formed.
       (os_ghost_mbx_are_well_formed
-       and then
+       and
```
Comment (jcdubois) [inline kernel/core/os.adb:733]: This all os_ghost_task_list_is_well_formed thing need to be worked out.

File: kernel/core/os.adb (hunk 1 above)
Comment (Blady-Com) [reply kernel/core/os.adb]: oops copy paste error, should have an additional "and" operator:
              os_ghost_current_task_is_ready and
              os_ghost_task_mbx_are_well_formed (dest_id),
As os_ghost_task_mbx_are_well_formed is just for one task, shouldn't be easier for the prover to match it directly rather than inside os_ghost_mbx_are_well_formed which is inside os_ghost_task_list_is_well_formed?
I don't really know, should not be. Something to confirm with Adacore.

File: kernel/core/include/os_arch.ads (hunk 2 above)
Comment (Blady-Com) [reply kernel/core/include/os_arch.ads:40]: I'm remembering some compiler issues when mixing aspect with pragma, could you try this:
   procedure os_arch_idle
     with Global => null,
     Post   => os.os_ghost_task_list_is_well_formed,
     Import => True,
     External_Name => "os_arch_idle",
     Convention => C;

File: kernel/core/include/os_arch.ads (hunk 5 above)
Comment (Blady-Com) [reply kernel/core/include/os_arch.ads:39]: "Global => null" is strictly concerning global variables access (in the Ada scope) which is closed to side effects.

File: kernel/core/include/os.ads (hunk 14 above)
Comment (Blady-Com) [reply kernel/core/include/os.ads]: You can express it as:
   type os_mbx_msg_t is range 0 .. 2 ** OS_MBX_MSG_SZ - 1;

File: kernel/core/os.adb (hunk 16 above)
Comment (Blady-Com) [reply kernel/core/os.adb:232]: I will have a try in a dedicated branch.

File: kernel/core/include/os_arch.ads (hunk 5 above)
Comment (jcdubois) [reply kernel/core/include/os_arch.ads:39]: wouldn't it be more correct then to use something like os_task_rw = os_task_rw'Old which is a lot better at expressing that the function has no side effect on the os_task_rw array.
We could add something similar on os_ghost_task_ready array.
But really it seems to me that this is what "Global => null" should mean implicitly.

File: kernel/core/include/os.ads (hunk 14 above)
Comment (jcdubois) [reply kernel/core/include/os.ads]: We certainly could ... But what does it buy us either functionally or on the proof front?

File: kernel/core/include/os_arch.ads (hunk 2 above)
Comment (jcdubois) [reply kernel/core/include/os_arch.ads:40]: We need to first resolve the comment above.
